File: js/연구소.js
const input = require('fs').readFileSync('/dev/stdin').toString().trim().split('\n');

const [n, m] = input[0].split(' ').map(Number);
const graph = input.slice(1, n+1).map((line) => line.split(' ').map(Number));

const dx = [-1,1,0,0];
const dy = [0,0,-1,1];
const answerList = [];

function bfs() {
    const queue = [];
    let tmpGraph = graph.map((row) => [...row]);

    for (let i=0; i<n; i++) {
        for (let j=0; j<m; j++) {
            if (tmpGraph[i][j] === 2) {
                queue.push([i,j]);
            }
        }
    }

    while (queue.length > 0) {
        let [x,y] = queue.shift();
        
        for (let i=0; i<4; i++) {
            let nx = x + dx[i];
            let ny = y + dy[i];

            if (nx >= 0 && nx < n && ny >= 0 && ny < m && tmpGraph[nx][ny] === 0) {
                tmpGraph[nx][ny] = 2;
                queue.push([nx,ny]);
            }
        }
    }
    
    const cnt = tmpGraph.flat().filter((v) => v === 0).length;
    answerList.push(cnt);
}

function makeWall(cnt) {
    if (cnt === 3) {
        bfs();
        return;
    }

    for (let i=0; i<n; i++) {
        for (let j=0; j<m; j++) {
            if (graph[i][j] === 0) {
                graph[i][j] = 1;
                makeWall(cnt+1);
                graph[i][j] = 0;
            }
        }
    }
}

makeWall(0);
console.log(Math.max(...answerList));

File: python/연구소.py
from collections import deque
import sys
import copy
input = sys.stdin.readline

n, m = tuple(map(int, input().split()))
graph = [
    list(map(int, input().split()))
    for _ in range(n)
]

dx = [-1,1,0,0]
dy = [0,0,-1,1]
cnt_list = []

def bfs():
    queue = deque()
    tmp_graph = copy.deepcopy(graph)

    for i in range(n):
        for j in range(m):
            # 바이러스라면
            if tmp_graph[i][j] == 2:
                queue.append((i,j))
    
    while queue:
        x,y = queue.popleft()

        for k in range(4):
            nx = x + dx[k]
            ny = y + dy[k]

            if 0 <= nx < n and 0 <= ny < m and tmp_graph[nx][ny] == 0:
                tmp_graph[nx][ny] = 2
                queue.append((nx,ny))

    cnt = 0
    for i in range(n):
        cnt += tmp_graph[i].count(0)
    
    cnt_list.append(cnt)

def makeWall(cnt):
    if cnt == 3:
        bfs()
        return
    
    for i in range(n):
        for j in range(m):
            if graph[i][j] == 0:
                graph[i][j] = 1
                makeWall(cnt+1)
                graph[i][j] = 0

makeWall(0)
print(max(cnt_list))